### PR TITLE
GitVersion: Remove support for obsolete versions

### PR DESCRIPTION
**Pull request by gerhardol**
Contributes to #8172 

## Proposed changes

GE currently officially supports Git from 2.19. Older than that the check reports red.
Unofficially, Git down to 1.7.1 or earlier is working though, there are checks that features are adapted to older versions.
I have previously advocated against removing support for older versions but now I have a reason:
To support WSL the solution gets more complex if older versions are still in the code.

This is not just creating commands. For instance the handling of encoding is spread to many more parts of the code (for something fixed 1.8.4).
Another example is that `git lfs clone` is obsolete and the checkbox can be removed.

To avoid that users use old Git and do not check Settings, a popup appears every time an old version is used.
This is also a preparation for WSL Git where other Git versions than the Windows version may be in use and you only want to be notified when using them.
(No point in trying to find all versions during startup or starting Settings.)

The officially recommended version as well as the officially oldest Git version (that GE reliably support) is bumped.
The officially oldest version is set to 2.25 (2020-01-13) a quite new version compared to current 2.19 (2018-09-18).
The popup as well as compatibility in the code is set to 2.17, not newer for two reasons:
* Git in WSL Ubuntu 18.04 uses 2.17 (I personally use this at work, even if I will migrate to 20.04 that comes with 2.25).
* This includes an example how compatibility is implemented for new commands. (There are two checks on 2.19 and one for 2.20)

## Test methodology 

Tests are updated. Some tests asserted that the correct Git version were used, other had compatibility code.
A test for RevisionReader was fixed, an obsolete test removed.

## Merge strategy

- Rebase merge (PR submitter must change the commit message for the last commit).
Some squash to be done.

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by gerhardol** [comment]
Squashed, removed two unused strings and changed the last supported version to 2.15.2 (there were no commands added in between 2.15.2 and 2.17).
Plan to rebase merge tomorrow, so this goes in before #9580

**Review by mstv — COMMENTED** [review]
minor comments, a little late

File: GitCommands/Git/GitVersion.cs [hunk 1]
```diff
@@ -1,31 +1,32 @@
...
+        /// <summary>
+        /// The oldest version with reasonable reliable support in GE.
+        /// Older than this version is red in settings.
+        /// </summary>
+        public static readonly GitVersion LastSupportedVersion = new("2.25.0");
+
+        /// <summary>
+        /// The oldest Git version without known incompatibilities.
+        /// </summary>
+        public static readonly GitVersion LastFailVersion = new("2.15.2");
```
[the remark this answers is not in the document]
Comment (gerhardol) [reply GitCommands/Git/GitVersion.cs:29]: `OldestKnownNotFailVersion`?

File: GitCommands/Git/GitVersion.cs [hunk 2]
```diff
@@ -37,6 +38,10 @@ public static GitVersion Current
                 {
                     string output = new Executable(AppSettings.GitCommand).GetOutput("--version");
                     _current = new GitVersion(output);
+                    if (_current < LastFailVersion)
+                    {
+                        MessageBox.Show(null, $"{_current} is lower than {LastSupportedVersion}. Some commands can fail.", "Unsupported Git version", MessageBoxButtons.OK, MessageBoxIcon.Error);
+                    }
```
[the remark this answers is not in the document]
Comment (gerhardol) [reply GitCommands/Git/GitVersion.cs:44]: Just did not bother with translation for this error message.
`LastSupportedVersion` is intentional, the detected version is the oldest, the oldest recommended is presented.
I can maybe make a follow up with these comments

File: GitCommands/Git/GitModule.cs [hunk 3]
```diff
@@ -3573,11 +3568,7 @@ public string GetFileText(ObjectId id, Encoding encoding)
 
             return messages.Select(cm =>
                 {
-                    int idx = cm.IndexOf('\n');
-                    string encodingName = cm.Substring(0, idx);
-                    cm = cm.Substring(idx + 1, cm.Length - idx - 1);
-                    cm = ReEncodeCommitMessage(cm, encodingName);
-                    return cm;
+                     return ReEncodeCommitMessage(cm);
                 });
```
[the remark this answers is not in the document]
Comment (gerhardol) [reply GitCommands/Git/GitModule.cs:3572]: seem ok

File: GitCommands/Git/GetAllChangedFilesOutputParser.cs [hunk 4]
```diff
@@ -28,14 +28,7 @@ public GetAllChangedFilesOutputParser(Func<IGitModule> getModule)
         /// <seealso href="https://git-scm.com/docs/git-status"/>
         public IReadOnlyList<GitItemStatus> Parse(string getAllChangedFilesCommandOutput)
         {
-            if (GitVersion.Current.SupportStatusPorcelainV2)
-            {
-                return GetAllChangedFilesFromString_v2(getAllChangedFilesCommandOutput);
-            }
-            else
-            {
-                return GetAllChangedFilesFromString_v1(getAllChangedFilesCommandOutput, false, StagedStatus.Unset);
```
[the remark this answers is not in the document]
Comment (gerhardol) [reply GitCommands/Git/GetAllChangedFilesOutputParser.cs:37]: yes

File: GitCommands/Git/GitVersion.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mstv) [reply GitCommands/Git/GitVersion.cs:29]: perhaps yes

File: GitCommands/Git/GitVersion.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mstv) [reply GitCommands/Git/GitVersion.cs:44]: > `LastSupportedVersion` is intentional,

Sensible. A regarding comment would make it clear for the reader.

**Pull request merged**